### PR TITLE
[Enhancement] .dev Python files updated to get better performance and syntax

### PR DESCRIPTION
**Pull request by Nourollah**
## Motivation

I started to read code details in MMSeg project for going to deep to project. Then I refactored some code part to better syntax

## Modification

I changed my latest commit in docs to explain details. Then I modified .dev directory Pythons file and setup.py.

### CODE REVIEW

File: .dev/gather_benchmark_train_results.py [hunk 1]
```diff
@@ -72,9 +72,9 @@ def parse_args():
                 print(f'log file error: {log_json_path}')
                 continue
 
-            differential_results = dict()
-            old_results = dict()
-            new_results = dict()
+            differential_results = {}
```
Comment (Nourollah) [inline .dev/gather_benchmark_train_results.py:75]: $ python3 -m timeit "x = {}"
20000000 loops, best of 5: 29.4 nsec per loop

File: .dev/gather_benchmark_train_results.py [hunk 2]
```diff
@@ -72,9 +72,9 @@ def parse_args():
                 print(f'log file error: {log_json_path}')
                 continue
 
-            differential_results = dict()
```
Comment (Nourollah) [inline .dev/gather_benchmark_train_results.py:75]: $ python3 -m timeit "x = dict()"
5000000 loops, best of 5: 69.8 nsec per loop

Comment (MeowZheng) [reply .dev/gather_benchmark_train_results.py:75]: Thanks for you improving mmseg code efficiency

File: .dev/gather_models.py [hunk 3]
```diff
@@ -50,25 +50,21 @@ def get_final_iter(config):
...
-            flag2 = (last_iter == iter_num - 50) or (last_iter == iter_num)
+            flag1 = 'aAcc' in log_line or log_line['mode'] == 'val'
+            flag2 = last_iter in [iter_num - 50, iter_num]
             if flag1 and flag2:
-                result_dict.update({
+                result_dict |= {
                     key: log_line[key]
                     for key in RESULTS_LUT if key in log_line
-                })
+                }
```
Comment (MeowZheng) [inline .dev/gather_models.py:66]: `|=` might unsupport dict

File: setup.py [hunk 4]
```diff
@@ -106,63 +99,45 @@ def add_mim_extension():
...
-        if platform.system() == 'Windows':
-            # set `copy` mode here since symlink fails on Windows.
-            mode = 'copy'
-        else:
-            mode = 'symlink'
-    elif 'sdist' in sys.argv or 'bdist_wheel' in sys.argv or \
-            platform.system() == 'Windows':
-        # installed by `pip install .`
-        # or create source distribution by `python setup.py sdist`
-        # set `copy` mode here since symlink fails with WinError on Windows.
```
Comment (MeowZheng) [inline setup.py:122]: I suggest these comments should be kept as a memoir of `mode=copy`

File: setup.py [hunk 5]
```diff
@@ -106,63 +99,45 @@ def add_mim_extension():
...
                 shutil.rmtree(tar_path)
-
             if mode == 'symlink':
                 src_relpath = osp.relpath(src_path, osp.dirname(tar_path))
                 try:
                     os.symlink(src_relpath, tar_path)
                 except OSError:
-                    # Creating a symbolic link on windows may raise an
-                    # `OSError: [WinError 1314]` due to privilege. If
-                    # the error happens, the src file will be copied
```
Comment (MeowZheng) [inline setup.py:149]: I suggest these comments should be kept

File: .dev/gather_models.py (hunk 3 above)
Comment (Nourollah) [reply .dev/gather_models.py:66]: That's supported, but you're right. It's not compatible with `Python < 3.9`. I return some changes to the previous version.

**Review by xiexinch — APPROVED** [review]

File: .dev/gather_models.py [hunk 6]
```diff
@@ -50,25 +50,21 @@ def get_final_iter(config):
...
+        for line in f:
             log_line = json.loads(line)
             if 'mode' not in log_line.keys():
                 continue
-
-            # When evaluation, the 'iter' of new log json is the evaluation
-            # steps on single gpu.
-            flag1 = ('aAcc' in log_line) or (log_line['mode'] == 'val')
-            flag2 = (last_iter == iter_num - 50) or (last_iter == iter_num)
+            flag1 = 'aAcc' in log_line or log_line['mode'] == 'val'
```
Comment (MeowZheng) [inline .dev/gather_models.py:60]:
```suggestion
            # When evaluation, the 'iter' of new log json is the evaluation
            # steps on single gpu.
            flag1 = 'aAcc' in log_line or log_line['mode'] == 'val'
```

**Review by MeowZheng — APPROVED** [review]

**Pull request merged**